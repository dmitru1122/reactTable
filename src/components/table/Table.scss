@import '../../theme/variables.scss';
.table {
  .table {
    &__row {
      cursor: pointer;
      &__cell {
        &--spinner {
          text-align: center !important;
        }
        &--short {
          cursor: initial;
          width: 15rem;
          @media screen and (max-width: $breakpoint-lg) {
            width: 10rem;
          }
        }
      }
    }
  }
}
