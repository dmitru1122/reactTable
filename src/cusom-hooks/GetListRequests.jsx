import { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import actionTypes from '../redux/actions/actionTypes';

export function useGetRequestData() {
  const requestsData = useSelector((state) => state.requestData);
  const lastUpdate = useSelector((state) => state.lastUpdate);
  const dispatch = useDispatch();
  const [data, setData] = useState(null);

  useEffect(() => {
    if (requestsData?.headerData) {
      setData(requestsData);
    } else {
      dispatch({ type: actionTypes.LOAD_DATA });
    }
  }, [lastUpdate, requestsData]);

  return data;
}

export const tti = 'sdfsd';
