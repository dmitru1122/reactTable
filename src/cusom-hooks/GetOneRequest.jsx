import { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import actionTypes from '../redux/actions/actionTypes';

export function useGetOneRequest(id) {
  const requestsData = useSelector((state) => state.fullRequestInfo);
  const dispatch = useDispatch();
  const [data, setData] = useState(null);

  useEffect(() => {
    if (requestsData[id]) {
      setData(requestsData[id]);
    } else {
      dispatch({ type: actionTypes.LOAD_ONE_REQUEST, id });
    }
  }, [requestsData]);

  return data;
}

export const test = 'dfsf';
