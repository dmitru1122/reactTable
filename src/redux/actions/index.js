import actionTypes from './actionTypes';

export function failure(error) {
  return {
    type: actionTypes.FAILURE,
    error,
  };
}

export function loadData() {
  return { type: actionTypes.LOAD_DATA };
}

export function loadDataSuccess(data, id) {
  return {
    type: actionTypes.LOAD_DATA_SUCCESS,
    data,
    id,
  };
}

export function addOneRequest(data) {
  return {
    type: actionTypes.ADD_ONE_REQUEST,
    data,
  };
}

// export function addOneRequestSuccess(data) {
//   return {
//     type: actionTypes.ADD_ONE_REQUEST_SUCCESS,
//     data,
//   };
// }
// export function addOneRequestFail() {
//   return {
//     type: actionTypes.ADD_ONE_REQUEST_FAIL,
//   };
// }

export function loadOneRequest(id) {
  return { type: actionTypes.LOAD_ONE_REQUEST, id };
}

export function loadOneRequestSuccess(data, id) {
  return {
    type: actionTypes.LOAD_ONE_REQUEST_SUCCESS,
    data,
    id,
  };
}

export function deleteOneRequest(id) {
  return { type: actionTypes.DELETE_ONE_REQUEST, id };
}
export function deleteOneRequestReset() {
  return { type: actionTypes.DELETE_ONE_REQUEST_RESET };
}

export function deleteOneRequestSuccess(id) {
  return { type: actionTypes.DELETE_ONE_REQUEST_SUCCESS, id };
}
export function deleteOneRequestFail() {
  return { type: actionTypes.DELETE_ONE_REQUEST_FAIL };
}

export function editOneRequest(data, id) {
  return { type: actionTypes.EDIT_ONE_REQUEST, data, id };
}
// export function editOneRequestSuccess(data, id) {
//   return { type: actionTypes.EDIT_ONE_REQUEST_SUCCESS, data, id };
// }
// export function editOneRequestFail() {
//   return { type: actionTypes.EDIT_ONE_REQUEST_FAIL };
// }
