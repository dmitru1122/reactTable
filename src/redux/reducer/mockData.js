export const initialState = {
  error: null,
  lastUpdate: 0,
  requestData: { headerData: null, rowData: null },
  deleteRequestStatus: 'waiting',
  fullRequestInfo: {},
  placeholderData: null,
};

export const rowData = [{ firstName: 'Vit', gender: 'male', id: '2', lastName: 'vab', purpose: 'Job proposition' }];
export const rowDataAfterFilter = [{ firstName: 'Vit', id: '2', purpose: 'Job proposition' }];
export const headerData = [
  { key: 'purpose', header: 'Purpose' },
  { key: 'firstName', header: 'First Name' },
];
export const fullHeaderData = [
  { key: 'purpose', header: 'Purpose' },
  { key: 'firstName', header: 'First Name' },
  { key: 'lastName', header: 'Last Name' },
  { key: 'gender', header: 'Purpose' },
];

export const failureMock = {
  data: {
    headerData,
    rowData: rowDataAfterFilter,
  },
  id: '2',
};

export const loadDataMock = {
  data: {
    headerData,
    rowData,
  },
  id: '2',
};
export const loadOneRequestMock = {
  data: {
    headerData: fullHeaderData,
    rowData,
  },
  id: '2',
};
export const loadSecondOneRequestMock = {
  data: {
    headerData: fullHeaderData,
    rowData: [{ firstName: 'Dima', gender: 'male', id: '3', lastName: 'Pashkevich', purpose: 'Job proposition' }],
  },
  id: '3',
};

export const addOneRequestDataMock = rowData;
//  {
//   firstName: 'Inna',
//   gender: 'male',
//   id: '',
//   lastName: 'test',
//   purpose: 'Job proposition',
// };

export const editOneRequestMock = {
  data: rowData,
  id: '2',
};
