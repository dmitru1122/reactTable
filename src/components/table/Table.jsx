import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';
import './Table.scss';
import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  DataTable,
  TableToolbar,
  TableToolbarContent,
  TableToolbarSearch,
  TableContainer,
  Table,
  TableHead,
  TableRow,
  TableHeader,
  TableBody,
  TableCell,
} from 'carbon-components-react';
import { deleteOneRequest, deleteOneRequestReset } from '../../redux/actions/index';
import Modal from '../modal/Modal';
import Notice from '../modal/Notice';
import SpinnerCS from '../spinner-cs/Spinner';

const propTypes = {
  headerData: PropTypes.arrayOf(PropTypes.shape({ key: PropTypes.string, header: PropTypes.string })),
  rowData: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      firstName: PropTypes.string,
      purpose: PropTypes.string,
    }),
  ),
};
const defaultProps = {
  headerData: [
    { key: 'purpose', header: 'Purpose' },
    { key: 'firstName', header: 'First Name' },
  ],
  rowData: [{ id: '', firstName: '', purpose: '' }],
};

function TableRequests(props) {
  const dispatch = useDispatch();
  const history = useHistory();
  const { rowData, headerData } = props;
  const [deleting, setDeleting] = useState({ status: 'beforeAction', id: 0 });
  const status = useSelector((state) => state.deleteRequestStatus);

  const openFullRequestInfo = (id) => {
    history.push(`/request/${id}`);
  };
  const handleClickDelete = (id) => {
    setDeleting({ status: 'working', id });
    dispatch(deleteOneRequest(id));
  };
  useEffect(() => {
    let reset = 0;

    if (status === 'resolve') {
      setDeleting({ status: 'success', id: 0 });
      reset = setTimeout(() => {
        dispatch(deleteOneRequestReset());
        setDeleting({ status: false, id: 0 });
      }, 5000);
    }
    if (status === 'reject') {
      setDeleting({ status: 'reject', id: 0 });
      reset = setTimeout(() => {
        dispatch(deleteOneRequestReset());
        setDeleting({ status: false, id: 0 });
      }, 5000);
    }
    return () => {
      clearTimeout(reset);
    };
  }, [status]);
  return (
    <div className='table'>
      <DataTable rows={rowData} headers={headerData} isSortable>
        {({ rows, headers, getHeaderProps, getTableProps, onInputChange, getBatchActionProps }) => (
          <TableContainer title='List of requests'>
            <TableToolbar>
              <TableToolbarContent>
                <TableToolbarSearch
                  tabIndex={getBatchActionProps().shouldShowBatchActions ? -1 : 0}
                  onChange={onInputChange}
                />
                {/* <Button
                  tabIndex={getBatchActionProps().shouldShowBatchActions ? -1 : 0}
                  onClick={() => handleClickEdit('addRequest', '0')}
                  renderIcon={Add}
                  kind='primary'>
                  Add new
                </Button> */}
                <Modal
                  type='add'
                  title='Add request'
                  buttonLabel='I have a component'
                  description='hi'
                  action='Delete'
                />
              </TableToolbarContent>
            </TableToolbar>
            <Table {...getTableProps()}>
              <TableHead>
                <TableRow>
                  {headers.map((header) => (
                    <TableHeader {...getHeaderProps({ header })}>{header.header}</TableHeader>
                  ))}
                  <TableHeader className='bx--table-sort-v2__ico'>Edit</TableHeader>
                </TableRow>
              </TableHead>
              <TableBody>
                {rows.map((row) => (
                  <TableRow key={row.id} className='table__row'>
                    {deleting.status === 'working' && deleting.id === row.id ? (
                      <td id='spin' className='table__row__cell--spinner' colSpan={row.cells.length + 1}>
                        <SpinnerCS />
                      </td>
                    ) : (
                      <>
                        {row.cells.map((cell) => (
                          <TableCell key={cell.id} onClick={() => openFullRequestInfo(row.id)}>
                            {cell.value}
                          </TableCell>
                        ))}
                        <TableCell className='table__row__cell--short'>
                          <Modal
                            type='delete'
                            title='Delete'
                            buttonLabel='I have a component'
                            description='Are you sure you want to go on?'
                            action='Delete'
                            continueAction={() => handleClickDelete(row.id)}
                          />

                          <Modal type='edit' title='Edit row' action='Delete' id={row.id} />
                        </TableCell>
                      </>
                    )}
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        )}
      </DataTable>
      <Notice isShowModal={deleting.status === 'success'} title='Warning' description='Request was deleted' />
      <Notice isShowModal={deleting.status === 'reject'} title='Error' description='Something was wrong' />
    </div>
  );
}

TableRequests.propTypes = propTypes;
TableRequests.defaultProps = defaultProps;

export default TableRequests;
