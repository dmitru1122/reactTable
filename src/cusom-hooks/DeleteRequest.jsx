import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { deleteOneRequest, deleteOneRequestReset } from '../redux/actions/index';

function useDeleteRequest() {
  const [id, setId] = useState();
  // const [status, setStatus] = useState('pending');
  const listRequests = useSelector((state) => state.requestData);
  const stat = useSelector((state) => state.deleteRequestStatus);
  const dispatch = useDispatch();
  console.log('rerender');
  listRequests.rowData.forEach((item) => {
    if (+item.id === +id && stat !== 'reject') {
      console.log('check pass, but it"s wrong');
      dispatch(deleteOneRequest(id));
    }
  });

  useEffect(() => {
    console.log(stat);
    // setStatus(stat);
    setTimeout(() => dispatch(deleteOneRequestReset()), 3000);
  }, [stat]);

  return { status: stat, setId };
}

export default useDeleteRequest;
