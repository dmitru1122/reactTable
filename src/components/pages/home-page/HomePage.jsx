import './HomePage.scss';
import useGetRequestData from '../../../cusom-hooks/GetListRequests';
import Table from '../../table/Table';
import SpinnerCs from '../../spinner-cs/Spinner';

function Home() {
  const listRequests = useGetRequestData();

  return (
    <main className='home'>
      {listRequests ? (
        <Table headerData={listRequests?.headerData} rowData={listRequests?.rowData} />
      ) : (
        <div className='d-flex align-items-center justify-content-center'>
          <SpinnerCs />
          {/* <div className='spinner-border text-primary' role='status'>
            <span className='visually-hidden'>Loading...</span>
          </div> */}
        </div>
      )}
    </main>
  );
}

export default Home;
