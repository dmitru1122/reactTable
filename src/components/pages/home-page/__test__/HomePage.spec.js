import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import HomePage from '../HomePage';
import Spinner from '../../../spinner-cs/Spinner';
import Table from '../../../table/Table';
import { useGetRequestData } from '../../../../cusom-hooks/GetListRequests';
import { headerData, rowData } from '../../../../redux/mockData';

jest.mock('../../../../cusom-hooks/GetListRequests');

const setUp = (store) =>
  mount(
    <Provider store={store}>
      <HomePage />
    </Provider>,
  );

describe('HomePage', () => {
  const initialState = { output: 10 };
  const mockStore = configureStore();
  const store = mockStore(initialState);
  let component;

  describe('loading', () => {
    beforeEach(() => {
      component = setUp(store);
    });

    it('should render component without crashing', () => {
      expect(component).toMatchSnapshot();
    });
    it('should render spinner ', () => {
      const element = <Spinner />;
      expect(component.contains(element)).toEqual(true);
    });
  });
  it('should render Table ', () => {
    useGetRequestData.mockReturnValue({ headerData, rowData });
    const wrapper = mount(
      <Provider store={store}>
        <HomePage />
      </Provider>,
    );

    const element = <Table headerData={headerData} rowData={rowData} />;
    expect(wrapper.contains(element)).toEqual(true);
  });
});
