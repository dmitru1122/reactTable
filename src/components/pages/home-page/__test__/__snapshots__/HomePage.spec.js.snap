// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage loading should render component without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Home>
    <main
      className="home"
    >
      <LoadTable
        data={null}
      >
        <div
          className="d-flex align-items-center justify-content-center"
        >
          <SpinnerCs>
            <div
              className="spinner-border text-primary"
              role="status"
            >
              <span
                className="visually-hidden"
              >
                Loading...
              </span>
            </div>
          </SpinnerCs>
        </div>
      </LoadTable>
    </main>
  </Home>
</Provider>
`;
